cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala

```diff
@@ -11,19 +11,54 @@ case class AccessUrlDownloader(accessUrl: AccessUrl, downloadLoc: String) extend
 
   def generateDownloadScript(): String = {
     val signedUrl = accessUrl.url
-    // TODO headers, refinements to retry strategy for certain 4xx statuses
-    s"""mkdir -p $$(dirname '$downloadLoc') && curl --location --retry 3 --retry-connrefused --retry-delay 10 --fail --output '$downloadLoc' '$signedUrl'"""
+    // TODO headers
+    s"""mkdir -p $$(dirname '$downloadLoc') && rm -f '$downloadLoc' && curl --silent --write-out '%{http_code}' --location --fail --output '$downloadLoc' '$signedUrl'"""
   }
 
-  override def download: IO[ExitCode] = {
-    // We don't want to log the unmasked signed URL here. On a PAPI backend this log will end up under the user's
-    // workspace bucket, but that bucket may have visibility different than the data referenced by the signed URL.
-    logger.info(s"Attempting to download data to '$downloadLoc' from access URL '${accessUrl.url.maskSensitiveUri}'.")
+  def returnCodeAndHttpStatus: IO[(Int, String)] = IO {
     val copyCommand = Seq("bash", "-c", generateDownloadScript())
     val copyProcess = Process(copyCommand)
+    val statusString = new StringBuilder()
+
+    val returnCode = copyProcess ! ProcessLogger({ s => statusString.append(s); () }, logger.underlying.error)
+    val httpStatus = statusString.toString().trim
+
+    (returnCode, httpStatus)
+  }
 
-    val returnCode = copyProcess ! ProcessLogger(logger.underlying.info, logger.underlying.error)
+  override def download: IO[DownloadResult] = {
+    // We don't want to log the unmasked signed URL here. On a PAPI backend this log will end up under the user's
+    // workspace bucket, but that bucket may have visibility different than the data referenced by the signed URL.
+    val masked = accessUrl.url.maskSensitiveUri
+    logger.info(s"Attempting to download data to '$downloadLoc' from access URL '$masked'.")
+
+    for {
+      rh <- returnCodeAndHttpStatus
+      (returnCode, httpStatus) = rh
+      ret = result(returnCode, httpStatus, masked)
+    } yield ret
+  }
 
-    IO(ExitCode(returnCode))
+  def result(returnCode: Int, httpStatus: String, maskedUrl: String): DownloadResult = {
+    (returnCode, httpStatus) match {
+      case (i, _) if i == 7 => // ECONNREFUSED
+        RetryableDownloadFailure(exitCode = ExitCode(i))
+      case (i, s) =>
+        try {
+          val intStatus = Integer.parseInt(s)
```

Comment:
TOL: This is fine, though I wonder if there's some HTTP status class that we can use. I'd imagine such a thing would be able to say what class of result it is (i.e. 2xx or 5xx) as well as check for specific codes.

Reply:
I found [this](https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/Response.Status.Family.html#familyOf-int-) in JEE 7 and [this](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/http/HttpStatusFamily.html#of-int-) in the AWS SDK and [this](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpStatus.Series.html#resolve-int-) in Spring but none of those are available here.